Tests/NIOPosixTests/RawSocketBootstrapTests.swift

```diff
@@ -0,0 +1,192 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import NIOCore
+@testable import NIOPosix
+
+extension NIOIPProtocol {
+    static let reservedForTesting = Self(rawValue: 253)
+}
+
+// lazily try's to create a raw socket and caches the error if it fails
+private let cachedRawSocketAPICheck = Result<Void, Error> {
+    let socket = try Socket(protocolFamily: .inet, type: .raw, protocolSubtype: .init(NIOIPProtocol.reservedForTesting), setNonBlocking: true)
+    try socket.close()
+}
+
+func XCTSkipIfUserHasNotEnoughRightsForRawSocketAPI(file: StaticString = #filePath, line: UInt = #line) throws {
+    do {
+        try cachedRawSocketAPICheck.get()
+    } catch let error as IOError where error.errnoCode == EPERM {
+        throw XCTSkip("Raw Socket API requires higher privileges: \(error)", file: file, line: line)
+    }
+}
+
+final class RawSocketBootstrapTests: XCTestCase {
+    func testBindWithRecevMmsg() throws {
+        try XCTSkipIfUserHasNotEnoughRightsForRawSocketAPI()
+        
+        let elg = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        defer { XCTAssertNoThrow(try elg.syncShutdownGracefully()) }
+        let channel = try RawSocketBootstrap(group: elg)
+            .channelInitializer {
+                $0.pipeline.addHandler(DatagramReadRecorder<ByteBuffer>(), name: "ByteReadRecorder")
+            }
+            .bind(host: "127.0.0.1", ipProtocol: .reservedForTesting).wait()
+        defer { XCTAssertNoThrow(try channel.close().wait()) }
+        try channel.configureForRecvMmsg(messageCount: 10)
+        let expectedMessages = (1...10).map { "Hello World \($0)" }
+        for message in expectedMessages {
+            _ = try channel.write(AddressedEnvelope(
+                remoteAddress: SocketAddress(ipAddress: "127.0.0.1", port: 0),
+                data: ByteBuffer(string: message)
+            ))
+        }
+        channel.flush()
+        
+        let receivedMessages = Set(try channel.waitForDatagrams(count: 10).map { envelop -> String in
+            var data = envelop.data
+            let header = try XCTUnwrap(IPv4Header(buffer: &data))
+            XCTAssertEqual(header.version, 4)
+            XCTAssertEqual(header.protocol, .reservedForTesting)
+            #if canImport(Darwin)
+            // On BSD the IP header will only contain the size of the ip packet body not the header.
+            // This is known bug which can't be fixed without breaking old apps which already workaround the issue
+            // like we do.
+            XCTAssertEqual(Int(header.totalLength), data.readableBytes)
+            // On BSD the checksum is always zero
+            XCTAssertEqual(header.headerChecksum, 0)
```

Comment:
I don't think we should make this platform specific. I think we should have a `bsdHeaderChecksum` and `bsdTotalLength` etc properties. And maybe maybe a `headerChecksumForPlatformRawSockets` or so properties that then (platform specifically) return the correct one.

Reply:
I thought about this too but couldn't come up with good names. These properties should return the correct value (as specified in [RFC-791](https://www.rfc-editor.org/rfc/rfc791#page-11)). The problem though is not only naming but also that the workaround (i.e. adding 20 to `totalLength`) requires knowing if we have just parsed the IPv4 packet from a `ByteBuffer` we got from a raw socket or not. If we create a `IPv4Header` ourselves e.g. for sending through the Raw Socket API, the totalLength will be correct and adding 20 will result in a wrong value.

The options I see are:
- store this information in the type and dynamically return the correct value
    - This works awful if we reuse the same value to e.g. echo a packet
- use different types for receiving and sending.
    - stores this information statically and we will need to explicitly convert the type when we want to echo a packet where we can get the `totalLength` fixed up
- use some rather long name which conveys that this is only for the receiving side e.g. `platformIndependetTotalLengthForRecivedPacket`
- add some documentation and leave it to the call site to handle it


I kind of want to use the "rather long name" variant but don't yet have name I like much.

Reply:
I also don't think we need to worry _too_ much about this, as this is testing code. It's ok if the parsing is a bit janky.

Reply:
I want to eventually make `IPv4Header` public API so we can still discuss this here. However, I will make this in a separate PR so it is not blocking `RawSocketBootstrap` to land.

Reply:
I'd go for some long names to explain this. And +1 for the public API, this should be reconciled with the already-existing IPv4/6 parsing code in swift-nio-extras.

Reply:
Added them in https://github.com/apple/swift-nio/pull/2320/commits/fabaf9b804afd085ff5d2c2f84ca8d2764e7b762